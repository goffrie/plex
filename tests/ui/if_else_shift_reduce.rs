//! This test demonstrates the classic if/else ambiguity in C,
//! which results in a shift-reduce conflict in the LR(1) grammar.

enum Token {
    If,
    Else,
    Var(i32),
    Semi,
    LParen,
    RParen,
}
type Span = ();
enum Expr {
    Var(i32),
}
enum Statement {
    If(Expr, Box<Statement>),
    IfElse(Expr, Box<Statement>, Box<Statement>),
}
#[allow(unused_imports)]
use self::Token::*;
plex::parser! {
    fn parse_(Token, Span);

    statements: Vec<Statement> {
        => vec![],
        statements[mut st] statement[e] Semi => {
            st.push(e);
            st
        }
    }

    statement: Statement {
        // The shift-reduce conflict applies here.
        // To resolve this conflict, add the following attribute, which effectively causes the ambiguity
        // to be resolved in favour of attaching the "else" to the inner "if" statement:
        // #[no_reduce(Else)]
        If LParen expr[head] RParen statement[then] => Statement::If(
            head,
            Box::new(then),
        ),
        If LParen expr[head] RParen statement[then] Else statement[else_] => Statement::IfElse(
            head,
            Box::new(then),
            Box::new(else_),
        ),
    }

    expr: Expr {
        Var(i) => Expr::Var(i),
    }
}

fn main() {
}
