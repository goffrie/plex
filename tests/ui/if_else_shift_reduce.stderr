error: shift-reduce conflict:
       state: statement -> If LParen expr RParen statement •
              statement -> If LParen expr RParen statement • Else statement
       token: Else
       path: statements If LParen expr RParen statement
  --> tests/ui/if_else_shift_reduce.rs:38:9
   |
38 |         If LParen expr[head] RParen statement[then] => Statement::If(
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
